docs/tutorials/writing_a_new_data_struct.md

```diff
@@ -0,0 +1,306 @@
+# Writing a New Data Struct
+
+ICU4X is a heavily data-driven library. Most new features or components will require pulling in data from an external source.
+
+This tutorial aims to help ICU4X contributors add new data to the data pipeline. It is recommended that readers review [data_pipeline.md](../design/data_pipeline.md) for additional theory behind the design decisions in the data provider.
+
+## Lifecycle of ICU4X Data
+
+It is important to understand the phases of life of ICU4X data as it makes its way from the data source, like CLDR, to the data struct used at runtime. The following flowchart shows the phases and how they connect:
+
+![Lifecycle of Data in ICU4X](../assets/data_lifecycle.svg)
+
+The following steps take place at build time:
+
+1. First, the source data file is obtained from an external source. Examples could include the CLDR JSON release or the Unicode Character Database.
+2. Second, we use a Serde definition to deserialize the source data file. This Serde implementation does not need to be optimized for performance.
+3. Third, we transform from the source data struct to the ICU4X runtime data struct. This step can be expensive, because it is normally run as an offline build step.
+4. Fourth, the ICU4X runtime data struct is serialized to either JSON, if debugging is important, or to a blob store, if being prepared for use in production.
+
+At runtime, only one step is performed: the data struct is deserialized from the JSON or blob emitted in step 4.
+
+When deserializing from the blob store, it is a design principle of ICU4X that no heap allocations will be required. We have many utilities and abstractions to help make this safe and easy.
+
+## Code Layout
+
+With a mental model of the lifecycle of data in ICU4X, we can discuss where to find the code that performs each step.
+
+### Data Structs
+
+The data struct definitions should live in the crate that uses them. By convention, the top-level module `provider` should contain the struct definitions. For example:
+
+- `icu::decimal::provider::DecimalSymbolsV1`
+- `icu::locale_canonicalizer::provider::LikelySubtagsV1`
+- `icu::uniset::provider::UnicodePropertyV1`
+
+In general, data structs should be annotated with `#[icu_provider::data_struct]`, and they should support *at least* `Debug`, `PartialEq`, `Clone`, `Default`, and Serde `Serialize` and `Deserialize`.
```

Comment:
Probably should also mention serde(borrow) and perhaps a bit about needing workarounds with Option (https://github.com/unicode-org/icu4x/pull/821https://github.com/unicode-org/icu4x/pull/821)

Reply:
I added more detail in *style_guide.md*, which has an existing section on this subject, and linked to it more prominently from the tutorial.